[USM]  Fix ssl_sock_by_ctx leak

amitslavin opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

Adds logic to handle connection closures without SSL_shutdown, relying solely on tcp_close. The solution involves creating a new map, `ssl_ctx_by_tuple`, to clean up the connection using the connection tuple (conn_tup).

### Motivation

The `ssl_sock_by_ctx` is leaking due to SSL connections being closed only via TCP close without calling SSL_SHUTDOWN.
This PR adds the ability to remove the entry from the map when a connection was closed.

### Describe how you validated your changes

- [x] Added a UT (TestSSLMapsCleanup) to validate the change. 
  I created a test that verifies the map is cleaned after the connection is closed. I also ran the test on the main branch to confirm that the map is not empty without the fix, and as expected, the test validates the improvement.
- [x] [Run](https://dddev.datadoghq.com/dashboard/gqu-2wz-asy/usm-performance-evaluation-20?fromUser=true&refresh_mode=paused&tpl_var_base_agent-env%5B0%5D=slavin-test-fix-base&tpl_var_client-service%5B0%5D=python-k6-client-grpc-tls&tpl_var_compare_agent-env%5B0%5D=slavin-test-fix-new&tpl_var_kube_cluster_name%5B0%5D=usm1&tpl_var_server-service%5B0%5D=python-grpcbin-tls&tpl_var_tls.library%5B0%5D=openssl&from_ts=1750593183043&to_ts=1750668730869&live=false) the branch on load test.
- [x] Run the branch on [staging](https://ddstaging.datadoghq.com/dashboard/kfn-zy2-t98/agent-system-probe-overview?fromUser=true&refresh_mode=paused&tpl_var_kube_cluster_name%5B0%5D=parent12&from_ts=1750166079253&to_ts=1750568461403&live=false) cluster.

I have created a [notebook](https://ddstaging.datadoghq.com/notebook/12190767/ssl-sock-by-ctx-map-leak?range=604800000&tpl_var_kube_cluster_name%5B0%5D=parent12&start=1749985045914&live=true) to track important metrics in order to validate that the leak was indeed resolved and to measure the impact of the change.

### Results

- A major improvement was observed in the ssl_sock_by_ctx leak, which can be clearly seen in the notebook.
- No negative impact on accuracy or performance was observed due to the change, both in the staging cluster and in the load test.
- No negative impact was observed in the eBPF internal telemetry, such as the user-mode map cleaner count, etc.
<img width="1394" alt="Screenshot 2025-06-23 at 13 39 32" src="https://github.com/user-attachments/assets/20801822-ce94-4b5f-a894-b6d024834795" />



<!--
Validate your changes before merge, ensuring that:

* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

### Possible Drawbacks / Trade-offs

### Additional Notes

Currently, this branch is based on the assumption that we are reverting PR [34981](https://github.com/DataDog/datadog-agent/pull/34981), as it handles the case where the map is not being used.

This solution was chosen to avoid changes that would significantly complicate the TLS code or introduce additional overhead.

TODO:

- [ ] Make sure that  PR [34981](https://github.com/DataDog/datadog-agent/pull/34981) is reverted before merging.


<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

amitslavin committed:
Revert "usm: exclude unused maps (#34981)"

This reverts commit a5ebedca1dba937cebf037fbbff4b664592af834.

amitslavin committed:
ssl cleanup from ssl main

amitslavin committed:
small changes

amitslavin committed:
fix test

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/fix1

amitslavin committed:
fix cr notes

guyarb commented on pkg/network/usm/ebpf_ssl.go:
```diff
@@ -666,6 +650,10 @@ func (o *sslProgram) cleanupDeadPids(alivePIDs map[uint32]struct{}) {
 			log.Debugf("SSL map %q cleanup error: %v", mapName, err)
 		}
+
+		if err := deleteDeadPidsInSSLCtxMap(o.ebpfManager, alivePIDs); err != nil {
+			log.Debugf("SSL map %q cleanup error: %v", sslCtxByPIDTGIDMap, err)
+		}
```
why is it inside the for-loop?

guyarb commented on pkg/network/usm/ebpf_ssl.go:
```diff
@@ -694,3 +682,34 @@ func deleteDeadPidsInMap(manager *manager.Manager, mapName string, alivePIDs map
 }
+
+// deleteDeadPidsInSSLCtxMap finds a map by name and deletes dead processes.
+func deleteDeadPidsInSSLCtxMap(manager *manager.Manager, alivePIDs map[uint32]struct{}) error {
```
what about the new map you've added `ssl_ctx_by_tuple`?

guyarb commented on pkg/network/usm/ebpf_ssl.go:
```diff
@@ -694,3 +682,34 @@ func deleteDeadPidsInMap(manager *manager.Manager, mapName string, alivePIDs map
 }
+
+// deleteDeadPidsInSSLCtxMap finds a map by name and deletes dead processes.
+func deleteDeadPidsInSSLCtxMap(manager *manager.Manager, alivePIDs map[uint32]struct{}) error {
```
why not using the MapCleaner interface?

amitslavin replied:
As we discussed, due to the way the cleanup is being done we will keep it like that

guyarb reviewed (changes requested):
reviewed

amitslavin committed:
create maps from ebpf ssl

amitslavin committed:
fix some cr notes

amitslavin committed:
fix cr notes and add new ut

amitslavin commented on pkg/network/usm/ebpf_ssl.go:
```diff
@@ -694,3 +682,34 @@ func deleteDeadPidsInMap(manager *manager.Manager, mapName string, alivePIDs map
+		pid := uint32(key >> 32)
+		if _, exists := alivePIDs[pid]; !exists {
+			_ = sslSockByCtxMapObj.Delete(unsafe.Pointer(&value))
+			keysToDelete = append(keysToDelete, key)
```
changed

guyarb commented on pkg/network/ebpf/c/protocols/tls/https.h:
```diff
@@ -327,6 +327,20 @@ static __always_inline void map_ssl_ctx_to_sock(struct sock *skp) {
+    bpf_map_update_with_telemetry(ssl_ctx_by_tuple, &ssl_sock.tup, &ssl_ctx, BPF_ANY);
+}
+
+static __always_inline void openssl_cleanup_maps(conn_tuple_t *tup) {
+    void **ssl_ctx_ptr = bpf_map_lookup_elem(&ssl_ctx_by_tuple, tup);
+    if (ssl_ctx_ptr) {
+        void *ssl_ctx = *ssl_ctx_ptr;
+        // Delete from tuple -> ctx map first
+        bpf_map_delete_elem(&ssl_ctx_by_tuple, tup);
+        // Then delete from original ctx -> sock_t map
+        if (ssl_ctx) {
+            bpf_map_delete_elem(&ssl_sock_by_ctx, &ssl_ctx);
+        }
+    }
 }
```
isn't that weird to make the implementation here, but the single usage elsewhere?

amitslavin replied:
Absolutely correct

amitslavin committed:
remove function

guyarb commented on pkg/network/usm/ebpf_ssl.go:
```diff
@@ -694,3 +714,57 @@ func deleteDeadPidsInMap(manager *manager.Manager, mapName string, alivePIDs map
+			sockKeysToDelete = append(sockKeysToDelete, sslCtxKey)
+
+			var sock http.SslSock
+			if err := sslSockByCtxMapObj.Lookup(unsafe.Pointer(&sslCtxKey), unsafe.Pointer(&sock)); err == nil {
+				tupleKeysToDelete = append(tupleKeysToDelete, sock.Tup)
+			}
```
another thing to consider, add the pid or tgid_pid to the key of that map, it will allow us easier deletion

guyarb committed:
Merge branch 'main' into amit.slavin/fix1

guyarb committed:
usm: openssl: Remove sslCtxByPIDTGID from sslPidKeyMaps list

The list holds map names to be cleaned with deleteDeadPidsInMap method, but we now have a dedicated method
to handle the cleanup in the map

guyarb committed:
usm: openssl: Initialize map cleaner

guyarb committed:
wip

brycekahle commented on pkg/network/usm/ebpf_main.go:
```diff
@@ -489,7 +489,11 @@ func (e *ebpfProgram) init(buf bytecode.AssetReader, options manager.Options) er
 				continue
 			}
-			options.ExcludedMaps = append(options.ExcludedMaps, m.Name)
+			// Unused maps still need to have a non-zero size
```
Why is `ExcludedMaps` not working for you?

brycekahle commented on pkg/network/usm/ebpf_ssl.go:
```diff
@@ -525,6 +541,36 @@ func (o *sslProgram) ConfigureOptions(options *manager.Options) {
+	}
+	o.sslSockByCtxMapObj = sslSockByCtxMapObj
+	cleaner2, err := ddebpf.NewMapCleaner[uint64, http.SslSock](sslSockByCtxMapObj, protocols.DefaultMapCleanerBatchSize/2, sslSockByCtxMap, UsmTLSAttacherName)
+	o.sslSockByCtxMapCleaner = cleaner2
```
missing `err` check here

guyarb replied:
True, it was missed by us
thanks for reviewing

amitslavin committed:
clean code and use 3 cleaners

amitslavin committed:
fix lint issue

amitslavin committed:
clean code

guyarb commented on pkg/network/usm/ebpf_ssl.go:
```diff
@@ -525,6 +538,38 @@ func (o *sslProgram) ConfigureOptions(options *manager.Options) {
 
 // PreStart is called before the start of the provided eBPF manager.
 func (o *sslProgram) PreStart() error {
+	sslCtxByPIDTGIDMapObj, _, err := o.ebpfManager.GetMap(sslCtxByPIDTGIDMap)
```
consider adding a simple function to remove the overhead here (get the map, check its error, create map cleaner, check its error)

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/fix1

amitslavin committed:
generic map cleaning function

amitslavin committed:
fix function

amitslavin committed:
use template function

amitslavin commented:
/merge

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/fix1

guyarb committed:
Merge branch 'main' into amit.slavin/fix1

guyarb committed:
Merge branch 'main' into amit.slavin/fix1
